docs/architecture/decisions/0004-generating-unique-ids.md

```diff
@@ -0,0 +1,47 @@
+
+We must define the format and method of generating these IDs. This can change for private beta, but should preferably be something we don’t need to change until then.
+
+Human-readable reference numbers are a better description rather than an IDs, as users may need to read them aloud over the phone or manually enter their reference number on the Use an LPA service. Therefore I will refer to them as reference numbers within this document. However they are not to be confused with the current LPA reference numbers in the format of 7000-0000-0000
+
+The reference numbers generated must:
+- Only use alphanumeric characters for readability purposes
+- Be unique to allow us to identify 1 LPA per reference number
+- Have a clear prefix to differentiate them from existing IDs. The Use an LPA service provides users with activation keys prefixed with a C- and allows users to create organisations codes to view their LPA prefixed with a V-. So this M- prefix is not only necessary to allow us to differentiate them from current LPAs, but also allow the user to differentiate their reference number(s) from Use an LPA activation keys/codes.
+
+
+For private beta (but not necessarily now) they also must:
+- Not be calculable (e.g. not a sequence) to avoid iteration based attacks.
+- Use the safe alphabet identified by Use an LPA `346789QWERTYUPADFGHJKLXCVBNM`. This is to avoid typographically similar letters such as i, l , and 1, making them more readable/usable for end users and avoid confusion.
+
+
+
+
+A problem that Use an LPA have experienced is words (particularly "bad" words) appearing within a reference number, ideally we would like a solution that prevents this.
+
+## Decision
+
+- The code will be 12 digits long to keep them consistent with current LPA UIDs & UAL activation key format and number of possible combinations. 
+- They will have an M- prefix to differentiate from existing reference numbers and Use an LPA activation keys (M for "Modernising"). For now (before private beta) we will prefix them with "MTEST-" to distinguish "test" cases.
+Eg: `M-3QT4-F65X-A7EJ`
+
+- To prevent users from being able to calculate/predict the reference numbers we will incorporate a random number generator such as the [math/rand Golang package](https://pkg.go.dev/math/rand) into the reference number generator. Seeding the pseudo-random number generator with the current time alone e.g. `source := rand.NewSource(time.Now().UnixNano())` would enable the reference number to be calculated, as each seed value will correspond to a sequence of generated values for a given random number generator. Therefore, if you provide the same seed twice, you get the same sequence of numbers twice, so seeding the current time in combination with another random variable e.g `source := rand.NewSource(rand.Int63() * time.Now().UnixNano())` will prevent prediction or calculation of reference numbers.
```

Comment:
Do avoid worrying about seeding, could we use [`rand.Int(..)` from crypto/rand](https://pkg.go.dev/crypto/rand#Int)?

Reply:
Yeah that looks a better to me :)